File: src/main/java/examples/TimerExample.java
package examples;

import com.netflix.spectator.api.Registry;
import com.netflix.spectator.api.Timer;

import javax.inject.Inject;
import java.util.concurrent.TimeUnit;


// #setup
public class TimerExample {

  private final Registry registry;
  private final Timer requestLatency;

  @Inject
  public TimerExample(Registry registry) {
    this.registry = registry;
    requestLatency = registry.timer("server.requestLatency");
  }
  // #setup

  // #using-lambda
  public Response handleUsingLambda(Request request) throws Exception {
    return requestLatency.record(() -> handleImpl(request));
  }
  // #using-lambda

  // #explicitly
  public Response handleExplicitly(Request request) {
    final long start = registry.clock().monotonicTime();
    try {
      return handleImpl(request);
    } finally {
      final long end = registry.clock().monotonicTime();
      requestLatency.record(end - start, TimeUnit.NANOSECONDS);
    }
  }
  // #explicitly

  private Response handleImpl(Request request) {
    // do something useful
    return new Response();
  }

  public static class Request {
  }

  public static class Response {
  }
}

File: src/main/python/timer-example.py

from spectator import GlobalRegistry

#setup
class TimerExample:

    def __init__(self, registry=GlobalRegistry):
        self._registry = registry
        self._requestLatency = registry.timer('server.requestLatency')
    #setup

    #using-with
    def handle_using_with(self, request):
        with self._requestLatency.stopwatch():
            return self._handle_impl(request)
    #using-with

    #explicitly
    def handle_explicitly(self, request):
        start = self._registry.clock().monotonic_time()
        try:
            return self._handle_impl(request)
        finally:
            end = self._registry.clock().monotonic_time()
            self._requestLatency.record(end - start)
    #explicitly

    def _handle_impl(self, request):
        # do something useful
        return None
